File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
    if (!head)
        return NULL;

    ListNode *temp = head;
    int count = 1;
    while (temp->next)
    {
        temp = temp->next;
        count++;
    }
    int res = count - n;
    if (res == 0)
    {
        head = head->next;
        return head;
    }
    temp = head;
    while (temp)
    {
        res = res - 1;
        if (res == 0)
        {
            break;
        }
        temp = temp->next;
    }
    temp->next = temp->next->next;
    return head;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
public ListNode removeNthFromEnd(ListNode head, int n) {
    if (head == null) return head;

    ListNode temp = head;
    int length = 0;

    while (temp != null) {
        temp = temp.next;
        length++;
    }

    int index = length - n;

    // 🔥 edge case: remove head
    if (index == 0) {
        return head.next;
    }

    temp = head;
    for (int i = 0; i < index - 1; i++) {
        temp = temp.next;
    }

    temp.next = temp.next.next;

    return head;
}
}